# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/canvas/WebGLContextGL.cpp
# Commit: 7e2a2b1b416f
# Full Hash: 7e2a2b1b416f1a64c6548aab29e92acc9554ae7d
# Author: Jeff Gilbert <jgilbert@mozilla.com>
# Date: 2020-01-09 03:40:18
# Regressor Bug: 1477756
# File Overlap Count: 1
# Description:
#   Bug 1477756 - Fix all webgl regression tests according to CI. r=handyman
#   
#   (This is a combination of 31 commits)
#   
#   * Fix Linux compilation.
# ==============================================================================

diff -r ccfa767dba64 -r 7e2a2b1b416f dom/canvas/WebGLContextGL.cpp
--- a/dom/canvas/WebGLContextGL.cpp	Wed Jan 08 22:19:16 2020 +0000
+++ b/dom/canvas/WebGLContextGL.cpp	Wed Jan 08 22:19:23 2020 +0000
@@ -1284,6 +1284,7 @@
   if (!locInfo) return;
 
   const auto& validationInfo = locInfo->info;
+  const auto& activeInfo = validationInfo.info;
   const auto& channels = validationInfo.channelsPerElem;
   const auto& pfn = validationInfo.pfn;
 
@@ -1291,14 +1292,23 @@
 
   const auto lengthInType = data.length() / sizeof(float);
   if (!lengthInType || lengthInType % channels != 0) {
-    const auto& activeInfo = validationInfo.info;
     GenerateError(LOCAL_GL_INVALID_VALUE,
-                  "`values` length (%u) must be a positive integer multiple of "
+                  "(uniform %s) `values` length (%u) must be a positive "
+                  "integer multiple of "
                   "size of %s.",
-                  lengthInType, EnumString(activeInfo.elemType).c_str());
+                  activeInfo.name.c_str(), lengthInType,
+                  EnumString(activeInfo.elemType).c_str());
     return;
   }
   const auto elemCount = lengthInType / channels;
+  if (elemCount > 1 && !validationInfo.isArray) {
+    GenerateError(
+        LOCAL_GL_INVALID_OPERATION,
+        "(uniform %s) `values` length (%u) must exactle match size of %s.",
+        activeInfo.name.c_str(), lengthInType,
+        EnumString(activeInfo.elemType).c_str());
+    return;
+  }
 
   // -
 
@@ -1362,11 +1372,11 @@
   funcScope.mBindFailureGuard = false;
 }
 
-void WebGLContext::ValidateProgram(const WebGLProgram& prog) {
+bool WebGLContext::ValidateProgram(const WebGLProgram& prog) const {
   const FuncScope funcScope(*this, "validateProgram");
-  if (IsContextLost()) return;
+  if (IsContextLost()) return false;
 
-  prog.ValidateProgram();
+  return prog.ValidateProgram();
 }
 
 RefPtr<WebGLFramebuffer> WebGLContext::CreateFramebuffer() {